### PR TITLE
Binary features projection fix

### PR DESCRIPTION
**Pull request by EugKar**
For the case of using binary features mapping matrix as projection layer, fixed potential gradients vanishing due to numerical errors caused by `exp` with huge values or `log` with small values.
Instead of converting unnormalized logits of binary features to log probabilities of outputs being zero or one via expressions which utilize `log` and `exp`, decoder is enforced to output explicit normalized log probabilities of ones and zeros (in other words, we got rid of sigmoid between decoder's outputs and projection layer). Additional regularizing term is added to loss to enforce decoder's outputs fit constraints for probability.

### CODE REVIEW

File: model_helper.py [hunk 1]
```diff
@@ -129,6 +129,14 @@ def compute_loss_sigmoid(logits, targets, final_sequence_length, target_sequence
 
     return loss
 
+def compute_log_probs_loss(outputs):
+    nfeatures = outputs.shape[-1] // 2
+    log_prob_ones = outputs[..., :nfeatures]
+    log_prob_zeros = outputs[..., nfeatures:]
```
Comment (akademi4eg) [inline model_helper.py]: To be on the safe side, I'd write `[..., nfeatures:2 * nfeatures]`

File: model_helper.py [hunk 2]
```diff
@@ -129,6 +129,14 @@ def compute_loss_sigmoid(logits, targets, final_sequence_length, target_sequence
 
     return loss
 
+def compute_log_probs_loss(outputs):
+    nfeatures = outputs.shape[-1] // 2
+    log_prob_ones = outputs[..., :nfeatures]
```
Comment (akademi4eg) [inline model_helper.py:138]: I think this should not be called `log_probs`, because it is not a normalized logit. Proper `Log(P)` is never positive.

File: model_helper.py [hunk 3]
```diff
@@ -129,6 +129,14 @@ def compute_loss_sigmoid(logits, targets, final_sequence_length, target_sequence
...
+    log_prob_zeros = outputs[..., nfeatures:]
+    norm_const = tf.stop_gradient(-(log_prob_ones + log_prob_zeros) / 2)
+    loss = tf.abs((tf.exp(log_prob_ones + norm_const) + tf.exp(log_prob_zeros + norm_const)) / tf.exp(norm_const) - 1)
```
Comment (akademi4eg) [inline model_helper.py:143]: This is equivalent to `1/p1 + 1/p0 -1`? So this loss attempts to keep probs near 0 and 1?

File: model_helper.py [hunk 4]
```diff
@@ -129,6 +129,14 @@ def compute_loss_sigmoid(logits, targets, final_sequence_length, target_sequence
...
+    loss += tf.nn.relu(log_prob_ones) + tf.nn.relu(log_prob_zeros)
```
Comment (akademi4eg) [inline model_helper.py:145]: This one tries to enforce `0 <= P <= 1` constraint? I think we might need to weight this part of loss with some regularization parameter.

Comment (EugKar) [reply model_helper.py:145]: Yes. I can add weight to the arguments, and set 1 by default, and later experiment with different values.

File: model_helper.py (hunk 2 above)
Comment (EugKar) [reply model_helper.py:138]: It is named so to emphasize the difference between a regular logit and this logit which is enforced to be normalized log probability by regularization term.

File: model_helper.py (hunk 3 above)
Comment (EugKar) [reply model_helper.py:143]: it is equivalent to enforce `p1 + p0 = 1`. Here `norm_const` is for numerical stability.

File: model_helper.py (hunk 2 above)
Comment (akademi4eg) [reply model_helper.py:138]: Ok, got it. I think some short comment is necessary hear. Otherwise at the first glance a reader would become puzzled why we add `relu(log(P))` to the loss. Just stating that this is a pseudoprobability and that this would eventually converge to a probability due to loss terms would be good.

File: model_helper.py (hunk 3 above)
Comment (akademi4eg) [reply model_helper.py:143]: Ah! I missed that stop gradient part! It is a bit counter-intuitive that the gradients correspond to `p1 + p0 - 1`, but the actual loss value corresponds to `1/p1 + 1/p0 - 1`. And with zero gradients you can have a loss equal to arbitrary number >= 3.

Comment (akademi4eg) [reply model_helper.py:143]: How about adding `tf.stop_gradient(-1/p1 - 1/p0 + p1 + p0)` to loss? This way loss and gradients would correspond to each other.

Comment (EugKar) [reply model_helper.py:143]: `stop_gradient` is applied to a normalization constant. It depends on `p1` and `p2` just to get proper value for each output. It is assumed to be constant, and is just used for numerical stability, and no gradients should pass through it. In other words, we might set it as `1e6`, for example, but it may not work for some outputs.

Comment (EugKar) [reply model_helper.py:143]: We multiply and divide equation `p1 + p0 = 1` by `e^norm_const`, so the actual loss remains `p1 + p0 - 1`.

Comment (akademi4eg) [reply model_helper.py:143]: You are right, my mistake.

**Review by akademi4eg — APPROVED** [review]
